src/org/traccar/database/NotificationManager.java

```diff
@@ -18,57 +19,70 @@
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
 import java.sql.SQLException;
-import java.util.Collection;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
-import java.util.concurrent.locks.ReadWriteLock;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import org.traccar.Context;
 import org.traccar.helper.Log;
 import org.traccar.model.Event;
 import org.traccar.model.Notification;
 import org.traccar.model.Position;
+import org.traccar.model.Typed;
 import org.traccar.notification.NotificationMail;
 import org.traccar.notification.NotificationSms;
 
-public class NotificationManager {
-
-    private final DataManager dataManager;
-
-    private final Map<Long, Set<Notification>> userNotifications = new HashMap<>();
-
-    private final ReadWriteLock notificationsLock = new ReentrantReadWriteLock();
+public class NotificationManager extends ExtendedObjectManager<Notification> {
 
     public NotificationManager(DataManager dataManager) {
-        this.dataManager = dataManager;
-        refresh();
+        super(dataManager, Notification.class);
+    }
+
+    private Set<Long> getEffectiveNotifications(long userId, long deviceId) {
```

Comment:
This function return ids of notifications that linked to user and either "always" or linked to device.
I decided that maps for every used objects are not necessary. Here is only one iteration for pair user-device.